test/algorithms/set_operations/intersection/intersection_areal_areal_linear.cpp

```diff
@@ -0,0 +1,318 @@
+struct closed_type<MultiPolygon, bg::multi_polygon_tag>
+{
+    typedef bg::model::multi_polygon
+        <
+            typename closed_type
+                <
+                    typename boost::range_value<MultiPolygon>::type
+                >::type
+        > type;
+};
+
+template <typename Areal1, typename Areal2, typename MultiLinestring>
+struct test_intersection_aal
+{
+    static inline void apply(std::string const& case_id,
+                             Areal1 const& areal1,
+                             Areal2 const& areal2,
+                             MultiLinestring const& expected1,
+                             MultiLinestring const& expected2)
+    {
+        typedef test_intersection_of_geometries
+            <
+                Areal1, Areal2, MultiLinestring
+            > tester;
+
+        tester::apply(areal1, areal2, expected1, expected2, case_id);
+
+        bool const is_open1 = (bg::closure<Areal1>::value == bg::open);
+        bool const is_open2 = (bg::closure<Areal2>::value == bg::open);
+
+        if (BOOST_GEOMETRY_CONDITION(is_open1 && is_open2))
+        {
+            typedef typename closed_type<Areal1>::type closed_areal1_type;
+            typedef typename closed_type<Areal2>::type closed_areal2_type;
+            closed_areal1_type closed_areal1;
+            closed_areal2_type closed_areal2;
+            bg::convert(areal1, closed_areal1);
+            bg::convert(areal2, closed_areal2);
+
+            typedef test_intersection_of_geometries
+                <
+                    closed_areal1_type, closed_areal2_type, MultiLinestring
+                > tester_of_closed;
+
+            std::string case_id_closed = case_id + "-closed";
+
+#ifdef BOOST_GEOMETRY_TEST_DEBUG
+            std::cout << "testing closed areal geometries..." << std::endl;
+#endif
+            tester_of_closed::apply(closed_areal1, closed_areal2,
+                                    expected1, expected2, case_id_closed);
+        }
+    }
+
+    static inline void apply(std::string const& case_id,
+                             Areal1 const& areal1,
+                             Areal2 const& areal2,
+                             MultiLinestring const& expected)
+    {
+        apply(case_id, areal1, areal2, expected, expected);
+    }
+};
+
+
+BOOST_AUTO_TEST_CASE( test_intersection_ring_ring_linestring )
+{
+#ifdef BOOST_GEOMETRY_TEST_DEBUG
+    std::cout << std::endl << std::endl << std::endl;
+    std::cout << "*** RING / RING / LINEAR INTERSECTION ***" << std::endl;
+    std::cout << std::endl;
+#endif
+    typedef open_ring_type OG;
+    typedef multi_linestring_type ML;
+
+    typedef test_intersection_aal<OG, OG, ML> tester;
+
+    tester::apply
+        ("r-r-01",
+         from_wkt<OG>("POLYGON((0 0,0 2,2 2,2 0))"),
+         from_wkt<OG>("POLYGON((2 1,2 4,4 4,4 0,1 0))"),
+         from_wkt<ML>("MULTILINESTRING((2 1,2 2),(2 0,1 0),(2 1,2 1))"),
+         from_wkt<ML>("MULTILINESTRING((2 2,2 1),(2 0,1 0),(2 1,2 1))")
```

Comment:
It's unclear to me why there are two expected WKT's which seem always the same.
Besides that, most probably a WKT of a multi linestring having the same linestrings but in a different order are also valid output. Is that checked?
I usually check only length (or area) - which is independent on order. Plus total number of points, for example. Admitted, different geometries could by chance produce the exactly same answer but chance is very low.